expression/extra/parser.py

```diff
@@ -0,0 +1,439 @@
+import string
+from typing import Any, Callable, Tuple, TypeVar, cast, overload
+
+from expression.collections import Block, block
+from expression.core import (
+    Error,
+    Nothing,
+    Ok,
+    Option,
+    Result,
+    Some,
+    curry,
+    curry_flipped,
+    match,
+    pipe,
+)
+
+_A = TypeVar("_A")
+_B = TypeVar("_B")
+_C = TypeVar("_C")
+_D = TypeVar("_D")
+
+
+ParseResult = Result[Tuple[_A, str], str]
+Parser = Callable[[str], ParseResult[_A]]
```

Comment:
@exyi This is a great suggestion! In the original F# code the parser is a single case union and it would be natural to extend it the same way as we have done for e.g option and result with "fluent" methods. Making it callable also avoids the explicit run function 🚀